File: java/src/main/java/com/myriad/wdl/model/statements/WdlScatter.java
package com.myriad.wdl.model.statements;

import com.myriad.wdl.model.definitions.WdlWorkflow.WdlWorkflowElement;
import com.myriad.wdl.model.expressions.WdlExpression;
import java.util.ArrayDeque;
import lombok.Getter;
import lombok.Setter;

/**
 * Scatter statement node.
 *
 * <p>A scatter iterates a bound name over a collection expression and evaluates a nested block of
 * workflow statements. See the WDL specification section "Scatter Statement".
 */
public final class WdlScatter implements WdlStatement, WdlWorkflowElement {
  @Getter @Setter private String name;
  @Getter @Setter private WdlExpression collection;
  private final ArrayDeque<WdlStatement> statements = new ArrayDeque<>();

  public WdlScatter() {}

  public WdlScatter(String name) {
    setName(name);
  }

  /** Returns the ordered nested statements inside the scatter body. */
  public ArrayDeque<WdlStatement> statements() {
    return statements;
  }

  @Override
  public ComponentType componentType() {
    return ComponentType.SCATTER;
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }
}

File: python/src/wdl_model/model/statements/wdl_scatter.py
"""Scatter statement nodes for the Python WDL object model."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from wdl_model.model.definitions import WdlWorkflowElement
from wdl_model.model.expressions import WdlExpression

from .wdl_statement import ComponentType, WdlStatement


@dataclass
class WdlScatter(WdlStatement, WdlWorkflowElement):
    """Scatter statement.

    A scatter iterates a bound name over a collection expression and evaluates a nested block of
    workflow statements.
    """

    name: str | None = None
    collection: WdlExpression | None = None
    _statements: deque[WdlStatement] = field(default_factory=deque)

    def statements(self) -> deque[WdlStatement]:
        """Return the ordered nested statements inside the scatter body."""
        return self._statements

    def componentType(self) -> ComponentType:
        """Return the statement category for traversal and validation dispatch."""
        return ComponentType.SCATTER
